rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1442,6 +1442,49 @@ impl SystemState {
             .time_out_messages(current_time, own_canister_id, local_canisters)
     }
 
+    /// Enqueues "deadline expired" references for all expired best-effort callbacks
+    /// without a response.
+    pub fn time_out_callbacks(
+        &mut self,
+        current_time: CoarseTime,
+        own_canister_id: &CanisterId,
+        local_canisters: &BTreeMap<CanisterId, CanisterState>,
+    ) {
+        if self.status == CanisterStatus::Stopped {
+            // Stopped canisters have no call context manager, so no callbacks.
+            return;
+        }
+
+        let aborted_or_paused_callback_id = self
+            .aborted_or_paused_response()
+            .map(|response| response.originator_reply_callback);
+
+        // Safe to unwrap because we just checked that the status is not `Stopped`.
+        let call_context_manager = call_context_manager_mut(&mut self.status).unwrap();
+
+        let expired_callbacks = call_context_manager
+            .expire_callbacks(current_time)
+            .collect::<Vec<_>>();
+        for callback_id in expired_callbacks {
+            if Some(callback_id) == aborted_or_paused_callback_id {
+                // This callback is already executing, don't produce a second response for it.
+                continue;
+            }
+
+            // Safe to unwrap because this is a callback ID we just got from the
+            // `CallContextManager`.
+            let callback = call_context_manager.callbacks().get(&callback_id).unwrap();
+            self.queues
+                .try_push_deadline_expired_input(
+                    callback_id,
+                    &callback.respondent,
+                    own_canister_id,
+                    local_canisters,
+                )
+                .unwrap();
```

Comment:
Is this unwrap here really safe? In particular, could there be cases where we interact with a dishonest subnet and one of the error conditions is triggered?

Even if it is safe, it might be worth thinking about raising a critical error, logging all the information and going ahead. It would seem to me that there is not much we win if we crash because we can't induct a timeout response.

Reply:
Same question, wouldn't a malicious sender send a response for an nonexistent request, for example?